### PR TITLE
🎉  trying chapter #1

### PR DESCRIPTION
**Pull request by capital-G**
### Solutions for Chapter 1

cc @vrom911 @chshersh 

First of all thanks for this amazing project!!!

The docs are really well written - its a great mixtuer of precise vocabularity and a good explaination of these constructs. 

## Bugs occured

https://github.com/kowainik/learn4haskell/compare/main...capital-G:solution-chapter1?expand=1#diff-540a810f208f97454f6e027d5774ce78f88d941006e6e9ef34c06c4b7390f759L531 seems off but maybe thats just me b/c its late.

Running the test suite results in where I am actually a bit lost

```
➜ make test-chapter1
cabal test doctest-chapter1 --enable-tests --test-show-details=direct
Build profile: -w ghc-8.10.1 -O1
In order, the following will be built (use -v for more details):
 - learn4haskell-0.0.0.0 (test:doctest-chapter1) (first run)
Preprocessing test suite 'doctest-chapter1' for learn4haskell-0.0.0.0..
Building test suite 'doctest-chapter1' for learn4haskell-0.0.0.0..
Running 1 test suites...
Test suite doctest-chapter1: RUNNING...
Examples: 38  Tried: 38  Errors: 0  Failures: 0
Test suite doctest-chapter1: PASS
Test suite logged to:
/Users/XXX/code/learn4haskell/dist-newstyle/build/x86_64-osx/ghc-8.10.1/learn4haskell-0.0.0.0/t/doctest-chapter1/test/learn4haskell-0.0.0.0-doctest-chapter1.log
1 of 1 test suites (1 of 1 test cases) passed.
cabal run learn4haskell-test --enable-tests -- -m "Chapter1"
Build profile: -w ghc-8.10.1 -O1
In order, the following will be built (use -v for more details):
 - happy-1.20.0 (exe:happy) (requires build)
 - quickcheck-io-0.2.0 (lib) (requires build)
 - pretty-show-1.10 (lib) (requires build)
 - hspec-core-2.7.4 (lib) (requires build)
 - hedgehog-1.0.4 (lib) (requires build)
 - hspec-2.7.4 (lib) (requires build)
 - hspec-hedgehog-0.0.1.2 (lib) (requires build)
 - learn4haskell-0.0.0.0 (test:learn4haskell-test) (first run)
Starting     quickcheck-io-0.2.0 (lib)
Starting     happy-1.20.0 (exe:happy)
Building     quickcheck-io-0.2.0 (lib)
Building     happy-1.20.0 (exe:happy)
Installing   quickcheck-io-0.2.0 (lib)
Completed    quickcheck-io-0.2.0 (lib)
Starting     hspec-core-2.7.4 (lib)
Building     hspec-core-2.7.4 (lib)
Installing   hspec-core-2.7.4 (lib)
Completed    hspec-core-2.7.4 (lib)
Starting     hspec-2.7.4 (lib)
Building     hspec-2.7.4 (lib)
Installing   hspec-2.7.4 (lib)
Completed    hspec-2.7.4 (lib)

Failed to build exe:happy from happy-1.20.0.
Build log ( /Users/XXX/.cabal/logs/ghc-8.10.1/hppy-1.20.0-9ac48e58.log ):
Configuring executable 'happy' for happy-1.20.0..
Preprocessing executable 'happy' for happy-1.20.0..
Building executable 'happy' for happy-1.20.0..
[ 1 of 19] Compiling AbsSyn           ( src/AbsSyn.lhs, dist/build/happy/happy-tmp/AbsSyn.o )
[ 2 of 19] Compiling GenUtils         ( src/GenUtils.lhs, dist/build/happy/happy-tmp/GenUtils.o )
[ 3 of 19] Compiling NameSet          ( src/NameSet.hs, dist/build/happy/happy-tmp/NameSet.o )
[ 4 of 19] Compiling ParamRules       ( src/ParamRules.hs, dist/build/happy/happy-tmp/ParamRules.o )
[ 5 of 19] Compiling ParseMonad       ( src/ParseMonad.hs, dist/build/happy/happy-tmp/ParseMonad.o )
[ 6 of 19] Compiling Lexer            ( src/Lexer.lhs, dist/build/happy/happy-tmp/Lexer.o )
[ 7 of 19] Compiling AttrGrammar      ( src/AttrGrammar.lhs, dist/build/happy/happy-tmp/AttrGrammar.o )
[ 8 of 19] Compiling AttrGrammarParser ( src/AttrGrammarParser.hs, dist/build/happy/happy-tmp/AttrGrammarParser.o )
[ 9 of 19] Compiling Grammar          ( src/Grammar.lhs, dist/build/happy/happy-tmp/Grammar.o )
[10 of 19] Compiling LALR             ( src/LALR.lhs, dist/build/happy/happy-tmp/LALR.o )
[11 of 19] Compiling First            ( src/First.lhs, dist/build/happy/happy-tmp/First.o )
[12 of 19] Compiling Parser           ( src/Parser.hs, dist/build/happy/happy-tmp/Parser.o )
[13 of 19] Compiling Paths_happy      ( dist/build/happy/autogen/Paths_happy.hs, dist/build/happy/happy-tmp/Paths_happy.o )
[14 of 19] Compiling Info             ( src/Info.lhs, dist/build/happy/happy-tmp/Info.o )
[15 of 19] Compiling PrettyGrammar    ( src/PrettyGrammar.hs, dist/build/happy/happy-tmp/PrettyGrammar.o )
[16 of 19] Compiling ProduceGLRCode   ( src/ProduceGLRCode.lhs, dist/build/happy/happy-tmp/ProduceGLRCode.o )
[17 of 19] Compiling Target           ( src/Target.lhs, dist/build/happy/happy-tmp/Target.o )
[18 of 19] Compiling ProduceCode      ( src/ProduceCode.lhs, dist/build/happy/happy-tmp/ProduceCode.o )
[19 of 19] Compiling Main             ( src/Main.lhs, dist/build/happy/happy-tmp/Main.o )
Linking dist/build/happy/happy ...
Undefined symbols for architecture x86_64:
  "___darwin_check_fd_set_overflow", referenced from:
      _awaitEvent in libHSrts.a(Select.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
`clang' failed in phase `Linker'. (Exit code: 1)
cabal: Failed to build exe:happy from happy-1.20.0 (which is required by
test:learn4haskell-test from learn4haskell-0.0.0.0). See the build log above
for details.

make: *** [test-chapter1] Error 1

```

### CODE REVIEW

**Comment by capital-G** [comment]
Build currently fails https://github.com/capital-G/learn4haskell/runs/1598373065

File: src/Chapter1.hs [hunk 1]
```diff
@@ -554,7 +556,13 @@ value after "=" where the condition is true.
...
 -}
 
-mid x y z = error "mid: not implemented!"
+mid :: Int -> Int -> Int -> Int
+mid x y z
+  | x <= y && y <= z = y
+  | z <= y && y <= x = y
+  | y <= x && x <= z = x
+  | z <= x && x <= y = x
+  | otherwise = z
```
Comment (chshersh) [inline src/Chapter1.hs:565]: Excellent implementation 🔝

File: src/Chapter1.hs [hunk 2]
```diff
@@ -568,7 +576,8 @@ True
 >>> isVowel 'x'
 False
 -}
-isVowel c = error "isVowel: not implemented!"
+isVowel :: Char -> Bool
+isVowel c = if c `elem` ['a', 'e', 'i', 'o', 'u'] then True else False
```
Comment (chshersh) [inline src/Chapter1.hs]: Nice one! 👍🏼 
As `elem` returns `Bool` itself, you can just return the result of this function, no need to use `if-then-else`

```suggestion
isVowel c = c `elem` ['a', 'e', 'i', 'o', 'u']
```

One note in here, that sometimes, `elem` could be slower than the explicit pattern matching. I remember there were some benchmarks on one particular case, that showed how moving to pattern matching on each case separately drastically decrease time 🐎

File: src/Chapter1.hs [hunk 3]
```diff
@@ -632,7 +641,11 @@ Try to introduce variables in this task (either with let-in or where) to avoid
 specifying complex expressions.
 -}
 
-sumLast2 n = error "sumLast2: Not implemented!"
+sumLast2 :: Int -> Int
+sumLast2 n = a + b
+  where
+    a = div ((abs n) `mod` 100) 10
+    b = (abs n) `mod` 10
```
Comment (chshersh) [inline src/Chapter1.hs:648]: That is a wonderful solution! 👏🏼  You correctly noticed that it is the `div` and `mod`, cool 😎 

One hint to make your solution even shorter: you can see that you use both:

```haskell
mod (abs n) 10
div (abs n) 10
```
The standard library has the [`divMod`](https://hackage.haskell.org/package/base-4.14.0.0/docs/Prelude.html#v:divMod) function, that actually combines inside both `div` and `mod`. And this is exactly what you use!.

So you could write it this way:

```haskell
(first, second) = divMod (abs n) 10
```

You can see how we could pattern match on the pair 🙂 And you will also use `abs` only once 👍🏻

File: src/Chapter1.hs [hunk 4]
```diff
@@ -653,7 +666,11 @@ You need to use recursion in this task. Feel free to return to it later, if you
 aren't ready for this boss yet!
 -}
 
-firstDigit n = error "firstDigit: Not implemented!"
+firstDigit :: Int -> Int
+firstDigit n
+  | div (abs n) 10 == 0 = abs n
+  | div (abs n) 10 > 0 = firstDigit (div (abs n) 10)
```
Comment (chshersh) [inline src/Chapter1.hs]: That is right 💯 

You can see that `abs n` is reused several times in this function. This usually means that you can move it into the `where` case and reuse everywhere 👍🏼

**Review by chshersh — APPROVED** [review]
@capital-G Great work on Chapter One tasks! 👏🏻 
And well done on the Advanced Task 💎 

Thanks for your kind words about our course! 😊  I left some comments with feedback and some suggestions, but solutions for this chapter already look great 👍🏻

**Comment by chshersh** [comment]
Hi @capital-G! The changes look good to me 👍🏻 
I think this PR is ready for merge 🏁

**Comment by capital-G** [comment]
> Hi @capital-G! The changes look good to me 👍🏻
> I think this PR is ready for merge 🏁

Thank you very much for the review! :)

**Pull request merged**